### ISSUE TITLE
Exception while running formatter

### ISSUE THREAD

**Issue by AndrewDryga** [rgreenjr/atom-elixir-formatter#24]
Received because of invalid Elixir code in the file which is similar to this one:
```
with :ok <- ok do
  foo = :bar
  foo =
else
  :error -> :ok
end
```

```
** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1    
    
    The following arguments were given to IO.chardata_to_string/1:
    
        # 1
        :stdin
    
    Attempted function clauses (showing 2 out of 2):
    
        def chardata_to_string(+string+) when -is_binary(string)-
        def chardata_to_string(+list+) when -is_list(list)-
    
    (elixir) lib/io.ex:461: IO.chardata_to_string/1
    (elixir) lib/path.ex:312: Path.relative_to/2
    (mix) lib/mix/tasks/format.ex:450: Mix.Tasks.Format.check!/1
    (mix) lib/mix/task.ex:316: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:79: Mix.CLI.run_task/2
```

**Comment by rgreenjr** [comment]
The plugin currently detects when `mix format` fails and shows the associated error message in a modal. Ideally, the error message will help users understand why formatting failed (although this error message isn't very informative).

I'm open to alternative approaches, however.

**Comment by AndrewDryga** [comment]
@rgreenjr the problem is that it does not show actual error but it fails to render one. You can compare exception above from the one you would get in CLI/during compilation and formatting via Code.format_string!/1:
```
iex(1)> with :ok <- :ok do
...(1)>     foo = 
...(1)>   end
** (SyntaxError) iex:3: syntax error before: 'end'

iex(1)> Code.format_string!("""
...(1)> with :ok <- :ok do
...(1)>     foo = 
...(1)>   end
...(1)> """)
** (SyntaxError) nofile:3: syntax error before: 'end'
    (elixir) lib/code.ex:538: Code.format_string!/2
    (stdlib) erl_eval.erl:680: :erl_eval.do_apply/6
    (iex) lib/iex/evaluator.ex:249: IEx.Evaluator.handle_eval/5
    (iex) lib/iex/evaluator.ex:229: IEx.Evaluator.do_eval/3
    (iex) lib/iex/evaluator.ex:207: IEx.Evaluator.eval/3
    (iex) lib/iex/evaluator.ex:94: IEx.Evaluator.loop/1
```

Even `mix format` task does it right:
```
$ mix format
mix format failed for file: first_responder.ex
** (SyntaxError) first_responder.ex:36: syntax error before: 'end'
    (elixir) lib/code.ex:538: Code.format_string!/2
    (mix) lib/mix/tasks/format.ex:397: Mix.Tasks.Format.format_file/2
    (elixir) lib/task/supervised.ex:89: Task.Supervised.do_apply/2
    (elixir) lib/task/supervised.ex:38: Task.Supervised.reply/5
    (stdlib) proc_lib.erl:249: :proc_lib.init_p_do_apply/3
```

**Comment by rgreenjr** [comment]
@AndrewDryga – It looks like that is the actual error generated by the Elixir formatter when run using `:stdin`, which is how Atom Elixir Formatter currently works.

If you run `mix format -` and type in the example followed by `CTRL-D` you get the same error:

```sh
$ mix format -
with :ok <- ok do
  foo = :bar
  foo =
else
  :error -> :ok
end

** (FunctionClauseError) no function clause matching in IO.chardata_to_string/1

    The following arguments were given to IO.chardata_to_string/1:

        # 1
        :stdin

    Attempted function clauses (showing 2 out of 2):

        def chardata_to_string(string) when is_binary(string)
        def chardata_to_string(list) when is_list(list)

    (elixir) lib/io.ex:461: IO.chardata_to_string/1
    (elixir) lib/path.ex:312: Path.relative_to/2
    (mix) lib/mix/tasks/format.ex:450: Mix.Tasks.Format.check!/1
    (mix) lib/mix/task.ex:316: Mix.Task.run_task/3
    (mix) lib/mix/cli.ex:79: Mix.CLI.run_task/2
```

Atom Elixir Formatter currently interacts with the Elixir formatter via stdin/stdout. However, the problems associated with this approach are starting to outweigh the benefits. I think I'm going to remove support for formatting text selections and unsaved buffers (which aren't used frequently) and change Atom Elixir Formatter to use the full path of the file being formatted. This will have the added benefit of significantly simplifying the code.

**Comment by rgreenjr** [comment]
@AndrewDryga I just published v2.0.0 which I believe resolves this issue. Please take a look and let me know if it works for you. Thanks.

**Comment by AndrewDryga** [comment]
@rgreenjr I tried it on a few cases where I got a generic error before and it looks like it got much better, thank you! 🎊
